RFDs/0001-job-worker-service.md

```diff
@@ -0,0 +1,468 @@
+
+  // Returns the list of all the jobs for the user
+  rpc ListJobs() returns (ListJobsResponse) {}
+}
+
+// StartJobRequest contains the information needed to start a new job.
+message StartJobRequest {
+  // The command to be executed.
+  string command = 1;
+  
+  // The arguments for the command.
+  repeated string args = 2;
+  
+  // The resource limits for the job.
+  ResourceLimits resource_limits = 3;
+}
+
+message JobId {
+  // The UUID of the job.
+  string id = 1;
+}
+
+// StartJobResponse contains the result of a job start request.
+message StartJobResponse {
+  // The UUID of the started job.
+  string id = 1;
+  
+  // The initial status of the job.
+  JobStatus status = 2;
+}
+
+// StopJobResponse contains the result of a job stop request.
+message StopJobResponse {
+  // The UUID of the stopped job.
+  string id = 1;
+  
+  // The final status of the job.
+  JobStatus status = 2;
+  
+  // The time when the job was started.
+  google.protobuf.Timestamp start_time = 3;
+  
+  // The time when the job was stopped.
+  google.protobuf.Timestamp stop_time = 4;
+}
+
+// JobStatusResponse contains the current status and resource usage of a job.
+message JobStatusResponse {
+  // The current status of the job.
+  JobStatus status = 1;
+  
+  // The current resource usage of the job.
+  ResourceUsage resource_usage = 2;
+}
+
+// JobOutputResponse contains a chunk of output data from a job.
+message JobOutputResponse {
+  // A chunk of output data from the job.
+  bytes data = 1;
+}
+
+// ResourceLimits defines the resource constraints for a job.
+message ResourceLimits {
+  // The number of CPU shares allocated to the job.
+  uint64 cpu_shares = 1;
+  
+  // The memory limit in bytes.
+  uint64 memory_limit_bytes = 2;
+  
+  // The I/O weight for prioritizing disk I/O.
+  uint64 io_weight = 3;
+}
+
+// ResourceUsage represents the current resource consumption of a job.
+message ResourceUsage {
+  // The current CPU usage.
+  uint64 cpu_usage = 1;
+  
+  // The current memory usage in bytes.
+  uint64 memory_usage = 2;
+  
+  // The current I/O usage.
+  uint64 io_usage = 3;
+}
+
+message ListJobsRequest {
+  // List of user ids to retrieve running jobs for
+  // Will return all if empty
+  repeated string user_ids = 1;
+}
+
+// List of jobs by user
+message ListJobsResponse {
+   message JobList {
+      repeated string job_ids = 1; 
+   }
+   map<string, JobList> jobs = 1;
+}
+
+// JobStatus represents the possible states of a job.
+enum JobStatus {
+  // The job is waiting to be executed.
+  PENDING = 0;
+  RUNNING = 1;
+  // The job has completed successfully.
+  COMPLETED = 2;
+  // The job has failed.
+  FAILED = 3;  
+  // The job was stopped by user request.
+  STOPPED = 4;
+}
+```
+
+#### Authentication and Authorization
+- Use mutual TLS (mTLS) for authentication
+- Use X.509 certificate extensions to store role information
+- Implement a role-based access control (RBAC) system with the following roles:
```

Comment:
Great choice to use the cert for authz, but RBAC with multiple roles is out-of-scope of the challenge requirements.

Reply:
Simplified it. Thanks!